src/main/java/se/liu/ida/hefquin/engine/federation/access/impl/FederationAccessManagerWithCache.java

```diff
@@ -0,0 +1,98 @@
+package se.liu.ida.hefquin.engine.federation.access.impl;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+
+import se.liu.ida.hefquin.engine.federation.BRTPFServer;
+import se.liu.ida.hefquin.engine.federation.Neo4jServer;
+import se.liu.ida.hefquin.engine.federation.SPARQLEndpoint;
+import se.liu.ida.hefquin.engine.federation.TPFServer;
+import se.liu.ida.hefquin.engine.federation.access.BRTPFRequest;
+import se.liu.ida.hefquin.engine.federation.access.CardinalityResponse;
+import se.liu.ida.hefquin.engine.federation.access.FederationAccessException;
+import se.liu.ida.hefquin.engine.federation.access.FederationAccessManager;
+import se.liu.ida.hefquin.engine.federation.access.Neo4jRequest;
+import se.liu.ida.hefquin.engine.federation.access.SPARQLRequest;
+import se.liu.ida.hefquin.engine.federation.access.SolMapsResponse;
+import se.liu.ida.hefquin.engine.federation.access.StringResponse;
+import se.liu.ida.hefquin.engine.federation.access.TPFRequest;
+import se.liu.ida.hefquin.engine.federation.access.TPFResponse;
+import se.liu.ida.hefquin.engine.query.TriplePattern;
+
+public class FederationAccessManagerWithCache implements FederationAccessManager 
+{
+	private FederationAccessManager fedAccMan;
+	private Map<TriplePattern, CompletableFuture<CardinalityResponse>> cacheMap;
```

Comment:
A few points here:
1. Both member variables should be `protected` in case we later want to create some subclass of this class, for whatever reason. Then, this subclass should have access to these member variables.
2. Both member variables should be `final`. This way, after they are initialized in the constructor, we cannot assign something else to them later. The idea is that we do not intend to later override the value of, say, `cacheMap` with another `Map` object because this way, we may lose the previous `Map` which, of course, is undesirable in the context of this class. The keyword `final` allows us to make this intention explicit and to get a compiler error in case we do anything unexpected later. So, using `final` is mainly a safety measure and it is generally a good practice to use it.
3. For member variables that do not need anything from the constructor to be initialized, it is preferable to include their initialization already in the declaration rather than in the constructor.

So, based on all of these points, these two lines should be changed as follows:
```suggestion
	protected final FederationAccessManager fedAccMan;
	protected final Map<TriplePattern, CompletableFuture<CardinalityResponse>> cacheMap = new HashMap<>();
```
By the way, notice that you can accept these suggestions directly here in the PR, in which case a commit is created automatically. The only thing that you need to be mindful about if you do so is that, if you later also want to do more code changes in your IDE, you have to do a `git pull` first in order to update your local branch to include these commits.